pkg/controller/raven/dns/dns_controller.go

```diff
@@ -0,0 +1,265 @@
+			fmt.Sprintf("The service %s/%s cluster IP is empty", utils.WorkingNamespace, utils.GatewayProxyInternalService))
+		return reconcile.Result{}, err
+	}
+
+	//4. update dns record
+	nodeList := new(corev1.NodeList)
+	err = r.Client.List(context.TODO(), nodeList, &client.ListOptions{})
+	if err != nil {
+		return reconcile.Result{Requeue: true}, fmt.Errorf("failed to list node, error %s", err.Error())
+	}
+	cm.Data[utils.ProxyNodesKey] = buildDNSRecords(nodeList, needProxy, svc.Spec.ClusterIP)
+	err = r.updateDNS(cm)
+	if err != nil {
+		return reconcile.Result{Requeue: true}, fmt.Errorf("failed to update configmap %s/%s, error %s",
+			cm.GetNamespace(), cm.GetName(), err.Error())
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *ReconcileDns) buildRavenDNSConfigMap() error {
+	cm := &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      utils.RavenProxyNodesConfig,
+			Namespace: utils.WorkingNamespace,
+		},
+		Data: map[string]string{
+			utils.ProxyNodesKey: "",
+		},
+	}
+	err := r.Client.Create(context.TODO(), cm, &client.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to create ConfigMap %s/%s, error %s", cm.GetNamespace(), cm.GetName(), err.Error())
+	}
+	return nil
+}
+
+func (r *ReconcileDns) getConfigMap(ctx context.Context, objectKey client.ObjectKey) (*corev1.ConfigMap, error) {
+	cm := corev1.ConfigMap{}
+	err := r.Client.Get(ctx, objectKey, &cm)
+	if err != nil {
+		return nil, err
+	}
+	return cm.DeepCopy(), nil
+}
+
+func (r *ReconcileDns) getService(ctx context.Context, objectKey client.ObjectKey) (*corev1.Service, error) {
+	svc := corev1.Service{}
+	err := r.Client.Get(ctx, objectKey, &svc)
+	if err != nil {
+		return nil, err
+	}
+	return svc.DeepCopy(), nil
+}
+
+func (r *ReconcileDns) updateDNS(cm *corev1.ConfigMap) error {
+	err := r.Client.Update(context.TODO(), cm, &client.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to update configmap %s/%s, %s", cm.GetNamespace(), cm.GetName(), err.Error())
+	}
+	return nil
+}
+
+func buildDNSRecords(nodeList *corev1.NodeList, needProxy bool, proxyIp string) string {
+	// record node name <-> ip address
+	var err error
+	dns := make([]string, 0, len(nodeList.Items))
+	for _, node := range nodeList.Items {
+		ip := proxyIp
+		if !needProxy {
+			ip, err = getHostIP(&node)
+			if err != nil {
+				klog.Errorf("failed to parse node address for %s, %s", node.Name, err.Error())
+				continue
+			}
+		}
+		dns = append(dns, fmt.Sprintf("%s\t%s", ip, node.Name))
+	}
+	sort.Strings(dns)
+	return strings.Join(dns, "\n")
+}
+
+func getHostIP(node *corev1.Node) (string, error) {
+	// get InternalIPs first and then ExternalIPs
+	allIPs := make([]net.IP, 0, len(node.Status.Addresses))
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == corev1.NodeInternalIP {
+			ip := net.ParseIP(addr.Address)
+			if ip != nil {
+				allIPs = append(allIPs, ip)
+				break
+			}
+		}
+	}
+	for _, addr := range node.Status.Addresses {
```

Comment:
reuse the same for loop, add else if to handle the same logic, reduce redundant code